### PR TITLE
Fixed hints for UNION.

### PR DESCRIPTION
**Pull request by igor-tkachev**

### CODE REVIEW

**Review by sdanyliv — APPROVED** [review]

File: Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs [hunk 1]
```diff
@@ -296,7 +296,7 @@ protected override void BuildInsertQuery(SqlStatement statement, SqlInsertClause
 				BuildStep = Step.HavingClause;    BuildHavingClause   (statement.SelectQuery);
 				BuildStep = Step.OrderByClause;   BuildOrderByClause  (statement.SelectQuery);
 				BuildStep = Step.OffsetLimit;     BuildOffsetLimit    (statement.SelectQuery);
-				BuildStep = Step.QueryExtensions; BuildQueryExtensions(statement);
+				BuildStep = Step.QueryExtensions; BuildQueryExtensions(statement, false, true);
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs:299]:
```suggestion
				BuildStep = Step.QueryExtensions; BuildQueryExtensions(statement, buildQueryHints: false, buildSubQueryHints: true);
```

File: Source/LinqToDB/DataProvider/Oracle/OracleSqlBuilderBase.cs [hunk 2]
```diff
@@ -711,9 +711,9 @@ protected override void FinalizeBuildQuery(SqlStatement statement)
 
 			if (statement.SqlQueryExtensions is not null && HintBuilder is not null)
 			{
-				if (HintBuilder.Length > 0 && HintBuilder[HintBuilder.Length - 1] != ' ')
+				if (HintBuilder.Length > 0 && HintBuilder[^1] != ' ')
 					HintBuilder.Append(' ');
-				BuildQueryExtensions(HintBuilder, statement.SqlQueryExtensions, null, " ", null);
+				BuildQueryExtensions(HintBuilder, statement.SqlQueryExtensions, null, " ", null, true, true);
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/Oracle/OracleSqlBuilderBase.cs:716]:
```suggestion
				BuildQueryExtensions(HintBuilder, statement.SqlQueryExtensions, null, " ", null, buildQueryHints: true, buildSubQueryHints: true);
```

File: Source/LinqToDB/SqlProvider/BasicSqlBuilder.cs [hunk 3]
```diff
@@ -361,7 +372,7 @@ protected virtual void BuildUpdateQuery(SqlStatement statement, SelectQuery sele
 			BuildStep = Step.OrderByClause;   BuildOrderByClause    (selectQuery);
 			BuildStep = Step.OffsetLimit;     BuildOffsetLimit      (selectQuery);
 			BuildStep = Step.Output;          BuildOutputSubclause  (statement.GetOutputClause());
-			BuildStep = Step.QueryExtensions; BuildQueryExtensions  (statement);
+			BuildStep = Step.QueryExtensions; BuildQueryExtensions  (statement, false, true);
```
Comment (viceroypenguin) [inline Source/LinqToDB/SqlProvider/BasicSqlBuilder.cs:375]:
```suggestion
			BuildStep = Step.QueryExtensions; BuildQueryExtensions  (statement, buildQueryHints: false, buildSubQueryHints: true);
```

File: Source/LinqToDB/SqlProvider/BasicSqlBuilder.cs [hunk 4]
```diff
@@ -417,7 +428,7 @@ protected virtual void BuildInsertQuery(SqlStatement statement, SqlInsertClause
 				BuildStep = Step.HavingClause;    BuildHavingClause   (statement.SelectQuery);
 				BuildStep = Step.OrderByClause;   BuildOrderByClause  (statement.SelectQuery);
 				BuildStep = Step.OffsetLimit;     BuildOffsetLimit    (statement.SelectQuery);
-				BuildStep = Step.QueryExtensions; BuildQueryExtensions(statement);
+				BuildStep = Step.QueryExtensions; BuildQueryExtensions(statement, false, true);
```
Comment (viceroypenguin) [inline Source/LinqToDB/SqlProvider/BasicSqlBuilder.cs:431]:
```suggestion
				BuildStep = Step.QueryExtensions; BuildQueryExtensions(statement, buildQueryHints: false, buildSubQueryHints: true);
```

**Review by viceroypenguin — DISMISSED** [review]

File: Tests/Linq/Extensions/QueryExtensionTests.cs [hunk 5]
```diff
@@ -121,5 +121,36 @@ select t
...
+			.AsSqlServer()
+				.WithReadUncommittedInScope()
+				.OptionRecompile()
+			.AsOracle()
+				.ParallelHint(2)
+			.AsAccess()
+				.WithOwnerAccessOption()
+			.AsMySql()
+				.MaxExecutionTimeHint(1000)
+			.AsPostgreSQL()
```
Comment (MaceWindu) [inline Tests/Linq/Extensions/QueryExtensionTests.cs:150]: is it correct? postgresql disabled for this test

**Comment by MaceWindu** [comment]
I wonder why we need wrapping subquery for second query (`ClickHouseTests.ClickHouseUnionTest`)

```sql
SELECT
	p.ID,
	p.TS
FROM
	ReplacingMergeTreeTable p FINAL
UNION DISTINCT
SELECT
	t2.ID,
	t2.TS
FROM
	(
		SELECT
			p_1.ID as ID,
			p_1.TS as TS
		FROM
			ReplacingMergeTreeTable p_1,
			(
				SELECT
					*
				FROM
					ReplacingMergeTreeTable t1
			) c_1 FINAL
	) t2
SETTINGS convert_query_to_cnf=false
```

**Review by MaceWindu — CHANGES_REQUESTED** [review]
subqueries with hints break subquery elimination which result in non-linear UNION. This will result in SQL errors in recursive CTEs with hints

**Comment by MaceWindu** [comment]
My bad, there is no issue with union, but still I wonder why we generate so many subqueries that look unnecessary

**Comment by MaceWindu** [comment]
Reviewed baselines once again, here are my questions:
- UnionTest: enable pgsql or remove pgsql hints
- PostgreSQLUnionTest: AsSubQuery moved joined table to subquery, but hint applied to main query which should put lock to both joined tables instead of one
- MySqlUnionTest: same issue with subquery and hint as with postgresql test. Also query hint location looks strange - I think it will apply to first subquery only

File: Tests/Linq/Extensions/QueryExtensionTests.cs (hunk 5 above)
Comment (igor-tkachev) [reply Tests/Linq/Extensions/QueryExtensionTests.cs]: fixed

**Comment by igor-tkachev** [comment]
> My bad, there is no issue with union, but still I wonder why we generate so many subqueries that look unnecessary

Is this related to this fix?

**Review by MaceWindu — APPROVED** [review]

**Pull request merged**